text/0000-template-invocation-context.md

```diff
@@ -0,0 +1,156 @@
+
+This suggests two possible approaches:
+
+a) A new named argument could be included, a function `_getInvocationStack` that could be accessed as follows:
+
+```js
+// functional helper
+// functional modifier
+// class-based helper
+
+let templateStack = named._getInvocationStack();
+
+// class-based modifier
+
+let templateStack = this.args.named._getInvocationStack();
+
+// component
+
+let templateStack = this.args._getInvocationStack();
+```
+
+In production, this method could return an empty object.
+
+a') Same as (a) but with a property getter instead of a function reference: e.g. `let templateStack = named._invocationStack;`
+
+b) A new debug function that can extract the stack from a hidden (symbol?) property stashed on the named/args object. It would look like this:
+
+```js
+import { getTemplateLocationInformation } from '@glimmer/debug';
+
+// functional helper
+// functional modifier
+// class-based helper
+
+let templateStack = getTemplateLocationInformation(named);
+
+// class-based modifier
+
+let templateStack = getTemplateLocationInformation(this.args.named);
+
+// component
+
+let templateStack = getTemplateLocationInformation(this.args);
+```
+
+In production, this method could return an empty object.
+
+In either case, we could consider a lint rule to make sure this is invoked within an `if (DEBUG) { ... }` block.
+
+### 3) How is the stack information is tracked by Glimmer and collected when requested via this API?
+
+I need help on this section. From discussing briefly with @rwjblue, I have learned that one of the precursors for implementation of something like this is to ensure that Ember's internal template compilation is aware of debug vs prod builds (today we compile templates with the same set of built-in AST plugins). This would be beneficial to reducing production build size, because the lack of differentiation in the current system has caused template size bloat because the compiler must always include debug information.
+
+## How we teach this
+
+This should be included in the API docs for helpers, modifiers and components with example code shown, along with example output. If a configuration option is included like the one posited under "Alternatives," this should be included in the guides under "Configuration > Debugging".
+
+When released, it should be included in the Ember.js Blog post for the release as well as highlighted in Ember Times. This could save existing Ember developers a lot of time and frustration.
+
+## Drawbacks
+
+APIs for helpers and modifiers are currently wonderfully simple and elegant. The addition this ugly debugging API compromises that design.
+
+The fact that that this will only be available in debug builds adds to complexity for consumers of the API and could potentially lead to errors in production if a developer does not realize this.
+
+## Alternatives
```

Comment:
Another possible alternative would be to provide a public API to emit the same hierarchical message that is used for the backtracking assertion (which is what the `on` and `fn` PR linked to above uses). This doesn’t solve the problem 100% (it wouldn’t directly provide access to the line and column info), but it’s fairly close and the plumbing is already largely available (still requires `owner` access though that might be mitigated by an argument as suggested in the api section). I _think_ this would likely be much easier to implement and could be improved over time to include line and column info when we have solved question 3 above (and in fact, would mean that **all** of Embers other template error messages like the backtracking rerender one, or on/fn mentioned above, would get that info also).

Reply:
@rwjblue Thanks for the quick feedback. I incorporated this and the other feedback you left into the text of the RFC. Feel free to suggest changes if your intention can be expressed more clearly than I captured.